docs: rewrite how-resolver-works.md to STE rules (#2046)
apache/maven-resolver: src/site/markdown/how-resolver-works.md

retrieved.
By default, Resolver operates with one local repository and zero or more remote
repositories.
The local repository is usually a directory on the local file system.
Remote repositories are usually HTTP servers.

*Resolving* is the process of finding an artifact from its coordinates and storing it locally.
It involves the following steps:

1. **Dependency graph collection** builds the dependency graph.
2. **Conflict resolution** removes conflicts, duplicates, and cycles from the
   graph. It produces the dependency tree.
3. **Flattening** transforms the tree into a list of artifacts. The list
   order represents the classpath order.
4. **Artifact resolution** finds each artifact in the flattened list in one of the available repositories. If necessary, it downloads
   the artifact from a remote repository and adds it to the local repository.

We call an artifact *resolvable* if it can be resolved from any available
repository.
The repository can be local or remote.
To make an artifact resolvable from the local repository, you install it.
To make an artifact resolvable from a remote repository, you deploy it.

> <aside>

Reviewer: Minor: `<aside>` is not used in any other Markdown file in the project's site. The Maven site skin likely has no CSS styling for it, so it would render identically to surrounding text with no visual distinction. A blockquote (`> ...`) would be safer and consistent with the rest of the docs.

Level order is the default.

This step operates entirely in memory.
The resolver does not download anything.

### Artifact Resolution

Artifact resolution checks to see if the binary artifact resource, most commonly a JAR file, is
in the local repository. If it isn't, Resolver downloads the file
from a remote repository and caches it in the local repository.

----

In general, Resolver builds the dependency graph and resolves conflicts in that graph in the same pass.
The name for this operation is *dependency collection*.

Resolver also usually resolves artifacts as it builds the flattened list.
The name for this combined step is *artifact resolution*.

When all the steps are performed together, the process is called *dependency
resolution*.

Yes, this is an unfortunate overload of terminology. 

The Resolver API reflects this terminology and offers methods for collection,
resolution, or both.

* `CollectResult collectDependencies(RepositorySystemSession session, CollectRequest request)` performs the dependency collection step. It builds the dependency graph and resolves conflicts in that graph before returning.

> * `List<ArtifactResult> resolveArtifacts(RepositorySystemSession session, Collection<ArtifactRequest> requests)` performs the artifact resolution step. It builds a flattened list and downloads artifacts before returning.

Reviewer: This description is still inaccurate. `resolveArtifacts` does NOT build a flattened list — it takes a pre-supplied `Collection<? extends ArtifactRequest>` and resolves the path for each artifact, downloading from remote repositories if necessary.

The source code at `DefaultRepositorySystem.java:308-334` confirms: `resolveArtifacts` simply validates requests and delegates to `artifactResolver.resolveArtifacts(session, requests)`. Flattening is performed by `doFlattenDependencyNodes` inside `resolveDependencies` (a different method).

The Javadoc says: _"Resolves the paths for a collection of artifacts."_

Suggested fix:
```suggestion
* `List<ArtifactResult> resolveArtifacts(RepositorySystemSession session, Collection<ArtifactRequest> requests)` performs the artifact resolution step. It resolves the path for each artifact in the provided collection, downloading from remote repositories if necessary.
```

under the License.
-->

Maven Artifact Resolver (formerly Aether) is a central piece of Maven.
This document explains how Resolver works internally.
It also explains the main concepts and components of Resolver.

Resolver alone is incomplete. It needs an application such as 
Maven to resolve versions and build effective models.
The Maven module `org.apache.maven:maven-resolver-provider` complements it with
models and implementations of missing components.

## Core Concepts

**Artifacts** and **repositories** are at the core of Resolver.

> An *artifact* is a binary resource with Maven coordinates.

Reviewer: "binary resource" is inaccurate — POM files are text/XML artifacts. The `Artifact` Javadoc says: "an artifact has identifying coordinates and optionally a file that denotes its data." Dropping "binary" would fix this.

```suggestion
An *artifact* is a resource identified by Maven coordinates.
```

Author: It is a binary resource. Maven loads it as bytes without processing or decoding text.

Usually it is a JAR file, but it can be anything as long as Maven coordinates can
address it.
The Maven coordinates are
`<groupId>:<artifactId>[:<extension>[:<classifier>]]:<version>`.
The default value of `extension` is `jar`.
The default value of `classifier` is an empty string.

A *repository* is a place where artifacts are stored and from where they can be
retrieved.
By default, Resolver operates with one local repository and zero or more remote
repositories.
The local repository is usually a directory on the local file system.
Remote repositories are usually HTTP servers.

*Resolving* is the process of finding

> an artifact from its coordinates and adding it to the Maven build.

Reviewer: "adding it to the Maven build" is inaccurate. Resolver is a standalone library — it resolves paths and downloads files. The calling application (Maven) is responsible for adding resolved artifacts to its build. The `resolveArtifact` Javadoc says: "Resolves the path for an artifact. The artifact will be downloaded to the local repository if necessary."

```suggestion
*Resolving* is the process of locating an artifact from its coordinates and obtaining its content.
```

It involves the following steps:

1. **Dependency graph collection** builds the dependency graph.
2. **Conflict resolution** removes conflicts, duplicates, and cycles from the
   graph. It produces the dependency tree.
3. **Flattening** transforms the tree into a list of artifacts. The list
   order represents the classpath order.
4. **Artifact resolution** finds each artifact in the flattened list in one of the available repositories. If necessary, it downloads
   the artifact from a remote repository and adds it to the local repository.

We call an artifact *resolvable* if it can be resolved from any available
repository.
The repository can be local or remote.
To make an artifact resolvable from the local repository, you install it.
To make an artifact resolvable from a remote repository, you deploy it.

> <aside>

Reviewer: The `<aside>` HTML tag is not used anywhere else in the project's site markdown. Doxia (Maven's site renderer) may not render it properly. A standard markdown blockquote would be safer and more consistent.

Author: This is not a block quote. It is an aside. We may not use this particular tag yet, but I think we do use pass-through HTML and Doxia does support it.

Level order is the default.

This step operates entirely in memory.
The resolver does not download anything.

### Artifact Resolution

Artifact resolution checks to see if the binary artifact resource, most commonly a JAR file, is
in the local repository. If it isn't, Resolver downloads the file
from a remote repository and caches it in the local repository.

----

In general, Resolver builds the dependency graph and resolves conflicts in that graph in the same pass.
The name for this operation is *dependency collection*.

Resolver also usually resolves artifacts as it builds the flattened list.
The name for this combined step is *artifact resolution*.

When all the steps are performed together, the process is called *dependency
resolution*.

Yes, this is an unfortunate overload of terminology. 

The Resolver API reflects this terminology and offers methods for collection,
resolution, or both.

* `CollectResult collectDependencies(RepositorySystemSession session, CollectRequest request)` performs the dependency collection step. It builds the dependency graph and resolves conflicts in that graph before returning.

> * `List<ArtifactResult> resolveArtifacts(RepositorySystemSession session, Collection<? extends ArtifactRequest> requests)` performs the artifact resolution step. It builds a flattened list and downloads artifacts before returning.

Reviewer: `resolveArtifacts` does not build a flattened list. Its signature takes a pre-supplied `Collection<? extends ArtifactRequest>` and resolves the path for each artifact in that collection. Flattening is done by `doFlattenDependencyNodes` inside `resolveDependencies`, not `resolveArtifacts`. The Javadoc says: "Resolves the paths for a collection of artifacts."

This has been flagged in previous reviews.

```suggestion
* `List<ArtifactResult> resolveArtifacts(RepositorySystemSession session, Collection<? extends ArtifactRequest> requests)` performs the artifact resolution step. It resolves the path for each artifact in the provided collection, downloading from remote repositories if necessary.
```

Author: Notice lines 148-149, or line 114 in the original. This might be wrong, it probably is, but it's not enough to just change it here. What else needs to be done?

under the License.
-->

Maven Artifact Resolver (formerly Aether) is a central piece of Maven.
This document explains how Resolver works internally.
It also explains the main concepts and components of Resolver.

Resolver alone is incomplete. It needs an application such as 
Maven to resolve versions and build effective models.
The Maven module `org.apache.maven:maven-resolver-provider` complements it with
models and implementations of missing components.

## Core Concepts

**Artifacts** and **repositories** are at the core of Resolver.

> An *artifact* is a binary resource with Maven coordinates.

Reviewer: "Binary resource" is inaccurate — POM files are text/XML and are valid artifacts. The `Artifact` interface Javadoc says: "an artifact has identifying coordinates and optionally a file that denotes its data." Consider:
```suggestion
An *artifact* is a resource identified by Maven coordinates and optionally backed by a file.
```

Author: The Artifact interface Javadoc is incorrect. This is correct. All text files are binary files. Not all binary files are text files.

retrieved.
By default, Resolver operates with one local repository and zero or more remote
repositories.
The local repository is usually a directory on the local file system.
Remote repositories are usually HTTP servers.

*Resolving* is the process of finding
an artifact from its coordinates and adding it to the Maven build.
It involves the following steps:

1. **Dependency graph collection** builds the dependency graph.
2. **Conflict resolution** removes conflicts, duplicates, and cycles from the
   graph. It produces the dependency tree.
3. **Flattening** transforms the tree into a list of artifacts. The list
   order represents the classpath order.
4. **Artifact resolution** finds each artifact in the flattened list in one of the available repositories. If necessary, it downloads
   the artifact from a remote repository and adds it to the local repository.

We call an artifact *resolvable* if it can be resolved from any available
repository.
The repository can be local or remote.
To make an artifact resolvable from the local repository, you install it.
To make an artifact resolvable from a remote repository, you deploy it.

<aside>

> Artifacts that are not in repositories also be resolved through extension points such as `WorkspaceReader`.

Reviewer: Grammar error — missing "can":
```suggestion
Artifacts that are not in repositories can also be resolved through extension points such as `WorkspaceReader`.
```

Maven does this when it exposes reactor projects, for example.
Normally you don't need to think about this.
</aside>

### Dependency Graph Collection

Collection is the first step.
Resolver adds a root artifact to a graph.
Then it adds the dependencies of the root artifact to the graph.
Then it adds the dependencies of the dependencies, and so on.
It stops when there are no more dependencies that haven't been added to the graph. 
The output of the collection step is a *dependency graph* known as the *dirty graph*.

> It can contain conflicts and duplicates.

Reviewer: `CollectResult` explicitly tracks `List<DependencyCycle>` (line 42, `getCycles()` at line 96). The dirty graph can also contain cycles:
```suggestion
It can contain conflicts, duplicates, and cycles.
```

During collection, only certain parts of the effective model are used, not the whole POM.

* `project/dependencies` defines the direct dependencies of a given node.
* `project/dependencyManagement/dependencies` defines the dependency management
  for subsequent nodes.
* `project/repositories` defines the repositories to be used on subsequent
  nodes.

In Resolver 1.x, `project/dependencyManagement` only defines versions of dependencies
for its own pom.xml. It does not affect the versions of the dependencies of the dependencies.
(*transitive dependencies*).

In Resolver 2.x, `project/dependencyManagement` does define versions of dependencies
for the transitive dependencies.

See also [common misconceptions](common-misconceptions.html).

### Conflict Resolution

Conflict resolution removes conflicts, duplicates, and cycles from the dependency graph.
The result is the **dependency tree**. This step operates entirely in memory.
The resolver does not download anything.

Resolver 2.x has two conflict resolution implementations.
The legacy implementation does multiple graph passes.
The faster path-based implementation does a single graph pass.

> TODO: do they give the same result?

Reviewer: This TODO should not appear in published documentation. Either answer the question and replace with a factual statement, or remove it.

Level order is the default.

This step operates entirely in memory.
The resolver does not download anything.

### Artifact Resolution

Artifact resolution checks to see if the binary artifact resource, most commonly a JAR file, is
in the local repository. If it isn't, Resolver downloads the file
from a remote repository and caches it in the local repository.

----

In general, Resolver builds the dependency graph and resolves conflicts in that graph in the same pass.
The name for this operation is *dependency collection*.

Resolver also usually resolves artifacts as it builds the flattened list.
The name for this combined step is *artifact resolution*.

When all the steps are performed together, the process is called *dependency
resolution*.

Yes, this is an unfortunate overload of terminology. 

The Resolver API reflects this terminology and offers methods for collection,
resolution, or both.

* `CollectResult collectDependencies(RepositorySystemSession session, CollectRequest request)` performs the dependency collection step. It builds the dependency graph and resolves conflicts in that graph before returning.

> * `List<ArtifactResult> resolveArtifacts(RepositorySystemSession session, Collection<? extends ArtifactRequest> requests)` performs the artifact resolution step. It builds a flattened list and downloads artifacts before returning.

Reviewer: This description is factually incorrect. `resolveArtifacts` takes a pre-supplied `Collection<? extends ArtifactRequest>` and resolves their paths — it does **not** build a flattened list. `DefaultRepositorySystem.resolveArtifacts` (lines 308-334) delegates directly to `artifactResolver.resolveArtifacts(session, requests)`. Flattening is done only inside `resolveDependencies` via `doFlattenDependencyNodes` (line 421-422).

Suggested:
```suggestion
* `List<ArtifactResult> resolveArtifacts(RepositorySystemSession session, Collection<? extends ArtifactRequest> requests)` performs only the artifact resolution step. It resolves the paths for a collection of artifacts, downloading them from remote repositories if necessary.
```

Author: Notice above where it says, "Resolver also usually resolves artifacts as it builds the flattened list.
The name for this combined step is *artifact resolution*." Is this also wrong? (it might be.) This was originally "The "flattening" and "artifact resolving" are also usually done together, and we use for those the term "artifact resolution"."
